### PR TITLE
[ValSet-Pref] Add `WithdrawDelegationRewards` message 

### PR DESCRIPTION
**Pull request by stackman27**
Part of : https://github.com/osmosis-labs/osmosis/issues/2579

## What is the purpose of the change

Adds new message `WithdrawDelegationRewards`. 
1. WithdrawDelegationRewards withdraws all the delegation rewards from the validator in the val-set.
2. Delegation reward is collected by the validator and in doing so, they can charge commission to the delegators.
3. Rewards are calculated per period, and is updated each time validator delegation changes.

## Brief Changelog
n/a 

## Testing and Verifying
Added msg_server test for to check functionality

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/valset-pref/msg_server_test.go [hunk 1]
```diff
@@ -292,3 +292,70 @@ func (suite *KeeperTestSuite) TestUnDelegateFromValidatorSet() {
...
+		delegator   sdk.AccAddress
+		coinToStake sdk.Coin
+		expectPass  bool
+	}{
+		{
+			name:        "Withdraw all rewards",
+			delegator:   sdk.AccAddress([]byte("addr1---------------")),
+			coinToStake: sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(20_000_000)), // delegate 20osmo
+			expectPass:  true,
+		},
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/valset-pref/msg_server_test.go:308]: Should test what happens if no valset set

File: x/valset-pref/types/expected_interfaces.go [hunk 2]
```diff
@@ -19,3 +19,10 @@ type StakingInterface interface {
 type BankKeeper interface {
 	GetBalance(ctx sdk.Context, addr sdk.AccAddress, denom string) sdk.Coin
 }
+
+type CommunityPoolKeeper interface {
```
Comment (ValarDragon) [inline x/valset-pref/types/expected_interfaces.go:23]:
```suggestion
type DistributionKeeper interface {
```

**Review by ValarDragon — APPROVED** [review]
logic LGTM, but needs more entries in table test case, at minimum for no valset set, no delegations, no valset explicitly set, one existing delegation

File: x/valset-pref/keeper_test.go [hunk 3]
```diff
@@ -41,6 +42,23 @@ func (suite *KeeperTestSuite) PrepareDelegateToValidatorSet() []types.ValidatorP
...
+
+	endingPeriod := suite.App.DistrKeeper.IncrementValidatorPeriod(ctx, validator)
+
+	delegation, found := suite.App.StakingKeeper.GetDelegation(ctx, delegator, valAddr)
+	suite.Require().True(found)
+
+	rewards := suite.App.DistrKeeper.CalculateDelegationRewards(ctx, validator, delegation, endingPeriod)
+
+	return rewards, validator
+}
```
[the remark this answers is not in the document]
Comment (stackman27) [reply x/valset-pref/keeper_test.go:60]: hmm this function is just being used in `msg_server_test.go` to check if delegationReward exists and returns existing reward. The reward validity is being checked in `msg_server_test.go` which is tdd. Func argument {val,  delegator} is checked in `msg_server_test.go`

File: x/valset-pref/types/expected_interfaces.go (hunk 2 above)
Comment (stackman27) [reply x/valset-pref/types/expected_interfaces.go:23]: Why for testing only?

**Comment by stackman27** [comment]
@ValarDragon @czarcas7ic added more tests coverage as you suggested

File: x/valset-pref/msg_server_test.go [hunk 4]
```diff
@@ -470,3 +470,94 @@ func (suite *KeeperTestSuite) TestRedelegateValidatorSet() {
...
+		coinToStake   sdk.Coin
+		setValSet     bool
+		setDelegation bool
+		expectPass    bool
+	}{
+		{
+			name:          "Withdraw all rewards with existing delegation",
+			delegator:     sdk.AccAddress([]byte("addr1---------------")),
+			coinToStake:   sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(20_000_000)), // delegate 20osmo
+			setValSet:     true,
```
Comment (ValarDragon) [inline x/valset-pref/msg_server_test.go:487]: Should be false, right?

Comment (ValarDragon) [reply x/valset-pref/msg_server_test.go:487]: or we need a test case with both this set, and it not set?

Comment (stackman27) [reply x/valset-pref/msg_server_test.go:487]: in this context "existing delegation" is "existing valset delegations" 

The WithdrawLogic also only check rewards from existing **valset delegations**

File: x/valset-pref/types/expected_interfaces.go (hunk 2 above)
Comment (ValarDragon) [reply x/valset-pref/types/expected_interfaces.go]: Its only used in testing? The main logic shouldn't use it?

Comment (stackman27) [reply x/valset-pref/types/expected_interfaces.go]: makes sense

File: x/valset-pref/msg_server_test.go (hunk 4 above)
Comment (ValarDragon) [reply x/valset-pref/msg_server_test.go:487]: Oh, I  would expect it to work s.t. if I have no valset explicitly set, it infers my valset is my existing delegation set, and withdraws from everyone via that

Comment (stackman27) [reply x/valset-pref/msg_server_test.go:487]: Yea i can get that to work, the only thing is that `GetDelegatorDelegations(ctx sdk.Context, delegator sdk.AccAddress, maxRetrieve uint16)` takes in uint16 and we might have to pass `math.MaxUint16` that could get a little expensive

File: x/valset-pref/types/expected_interfaces.go (hunk 2 above)
Comment (stackman27) [reply x/valset-pref/types/expected_interfaces.go]: We are using `WithdrawDelegationRewards(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) (sdk.Coins, error)` in code that falls under this interface

File: x/valset-pref/msg_server_test.go (hunk 4 above)
Comment (stackman27) [reply x/valset-pref/msg_server_test.go]: Just added withdraw existing delegation reward!

**Comment by stackman27** [comment]
@ValarDragon @czarcas7ic added test cases and logic for existing staking position that's not valset. lmk what you think

File: x/valset-pref/validator_set.go [hunk 5]
```diff
@@ -207,6 +208,84 @@ func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, e
...
+	delegator, err := sdk.AccAddressFromBech32(delegatorAddr)
+	if err != nil {
+		return err
+	}
+
+	// check if there is existing staking position that's not val-set
+	delegations := k.stakingKeeper.GetDelegatorDelegations(ctx, delegator, math.MaxUint16)
+
+	// get the existing validator set preference
+	existingSet, found := k.GetValidatorSetPreference(ctx, delegatorAddr)
```
Comment (ValarDragon) [inline x/valset-pref/validator_set.go:225]: I guess I'm confused -- I thought the module design was always such that `GetValidatorSetPreference` returns your existing staking distribution if you haven't set one.

Is that not the case or only in query logic?

Comment (stackman27) [reply x/valset-pref/validator_set.go:225]: Unfortunately that is not the case, `GetValidatorSetPreference` only returns val-set delegations

Comment (ValarDragon) [reply x/valset-pref/validator_set.go:225]: ok, imo we need to change that (which I thought was the initial design for valset pref)

Comment (stackman27) [reply x/valset-pref/validator_set.go:225]: for sure, it wont be a massive change tbh. Logic will be something like. 
```
func() GetValidatorSetPreference {
valsetDels, vsFound := getvalsetDels(...) 
 if !vsFound {
   existingstakingDels, stFound := getstakingDels(...) 
   if !stFound {
    return err ("No delegations") 
 }
return existingstakingDels 
} 
return valsetDels 
}
```

What about cases where we need both existing valset and staking position? Do you think this function should handle that too?

Comment (ValarDragon) [reply x/valset-pref/validator_set.go:225]: 1) if a valset is set return that. (ignore staking position, that may diverge)
2) You can derive the weight on the fly based on ratio of staked amounts

Comment (ValarDragon) [reply x/valset-pref/validator_set.go:225]: I'm down to merge this PR, but lets do the above as a release blocking feature.

Can you make an issue? Then we can merge this PR

Comment (stackman27) [reply x/valset-pref/validator_set.go:225]: added issue https://github.com/osmosis-labs/osmosis/issues/3848

**Review by czarcas7ic — APPROVED** [review]
LGTM, now that issue has been created, I will merge this

File: x/valset-pref/keeper_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/valset-pref/keeper_test.go]: This comment was placed in the wrong location, but it looks like you addressed it in later commits

**Pull request merged**